Fix part of #5: Abstraction on top of Oppia GAE for Topic Management System

rt4914 opened the pull request:
## Explanation
The PR contains data module, necessary libraries for API calls and Topic Management System models and retrofit instance.

## Libraries Used
1. **Retrofit:** implementation 'com.squareup.retrofit2:retrofit:2.5.0'
This library is responsible for making API calls to Oppia backend
2. **Retrofit GSON Convertor:** implementation 'com.squareup.retrofit2:converter-gson:2.5.0'
This library is a temporary fix for XSSI_PREFIX. This library can be removed once we use interceptor which will remove the XSSI_PREFIX from the response.
3. **Retrofit Moshi Convertor:** implementation 'com.squareup.retrofit2:converter-moshi:2.5.0'
This library is the actual deserializer that we are using over GSON.
4. **Moshi Kotlin Support:** implementation"com.squareup.moshi:moshi-kotlin:1.8.0"
Moshi Kotlin support library for using Moshi in Kotlin.

## Issues
The models in this PR does not have custom field names because @Json in Moshi which is similar to @SerializedName in GSON is not working with Kotlin Moshi.

## Checklist
Work items:

- [x] Identify among all existing GAE endpoints (get/post), which we can use during the prototype
- [x] Loosely identify data missing from the above endpoints (this bucket is expected to be small/non-existent), and identify which endpoints have data which needs to be changed for the Android app
- [x] Specify the Retrofit interfaces to represent all of the identified endpoints (simple interfaces/no annotations or return types needed beyond pseudocode)
- [x] Identify** what data (via pseudocode data structures/YAML/JSON/etc) we will be passing along from the backend through these endpoints
- [ ] Identify how error handling will work (e.g. what Retrofit does), whether RPC retry is supported & how it is/can be configured, etc.
- [ ] Determine the testing strategy for the GAE endpoints for downstream app components

Commit:
Model and retrofit instance for TopicService

Commit:
Model and retrofit instance for TopicService

Commit:
Removed Gson and solved issue of custom name fields in model

Commit:
Removed Gson and solved issue of custom name fields in model

Commit:
cleaning of code

Commit:
cleaning of code

rt4914 commented:
@BenHenning I have removed Retrofit-Gson-Converter library and made introduces an interceptor to modify the json response to remove XSSI_PREFIX. Now everything is working perfectly without using any extra libraries.

rt4914 commented:
@BenHenning PTAL

seanlip commented on .idea/dictionaries/talesra.xml:
```diff
@@ -0,0 +1,8 @@
+      <w>xssi</w>
+    </words>
+  </dictionary>
+</component>
```
Please add newline before EOF, here and elsewhere.

You can check the "Files Changed" tab to see which files need it.

rt4914 replied:
@seanlip I have added new line at the end in the files which I have written and some files are automatically generated in which EOF is not there. So should I change those generated files too? Like the one which you have mentioned.

seanlip replied:
Can you find a way to generate them so that a newline is added at EOF?

seanlip replied:
(In any case, the short answer to your question is yes. Though it may be worth checking with @BenHenning whether we should be checking in generated files into source control.)

rt4914 replied:
@seanlip I have activated this from settings in android studio.

seanlip replied:
I don't understand, sorry. What have you activated?

Also, please check your spelling in .gitignore. That might be source of some of the issues.

rt4914 replied:
> Can you find a way to generate them so that a newline is added at EOF?

In Android Studio -> Settings -> Editor Settings we have an option to select "Add newline before EOF" in generated files. That way even the generated files will have a new line before EOF.

rt4914 replied:
This has been solved.

seanlip commented on .gitignore:
```diff
@@ -7,6 +7,7 @@
 /.idea/workspace.xml
 /.idea/navEditor.xml
 /.idea/assetWizardSettings.xml
+/.idea/dictionaires/talesra.xml
```
I don't think we should have per-user details in .gitignore. @BenHenning?

rt4914 replied:
Solved from code side. Removed talesra.xml and introduced a common project-level dictionary

seanlip commented on data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt:
```diff
@@ -0,0 +1,34 @@
+
+import java.io.IOException
+
+//TODO: Transfer this XSSI_PREFIX to a constant file
```
You might as well do this now, here and elsewhere? Try to avoid leaving TODOs behind you in the codebase.

rt4914 replied:
Actually these TODOs are just a small part of a larger PR that we will have to do at later stage. Thus cannot solve this right now.

seanlip commented on data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt:
```diff
@@ -0,0 +1,34 @@
+        return response.newBuilder().body(body).build()
+      }
+    } else if (response.code() == 403) {
+      //TODO: Manage other network errors here
```
All TODOs should be formatted as follows:

``` TODO(#...): ...```

where the thing in parentheses is an issue number that describes what needs to be done in detail.

rt4914 replied:
These TODOs do not have an issue number yet. Should I create an issue and than change the TODO accordingly?

seanlip replied:
Yes, please.

rt4914 replied:
2 TODOs removed and made changes to remaining TODO

Commit:
EOF changes in files

Commit:
EOF changes in files

Commit:
cleaning of code

Commit:
cleaning of code

Commit:
removed extra lines from code

Commit:
removed extra lines from code

rt4914 commented:
Hi @seanlip , I generally put extra lines in between code sections for more clear code vision (Check Topic.kt file for the that), so should I continue doing that or not?

BenHenning commented on .idea/dictionaries/talesra.xml:
```diff
@@ -0,0 +1,8 @@
+<component name="ProjectDictionaryState">
```
Since this is specifically for you, maybe add it to .gitignore or, if possible, add these words to a shared dictionary so that everyone can benefit from them.

rt4914 replied:
Added shared project dictionary. Though currently IDE only add words to the <username> based dictionary, therefore developers will have to add words manually in this.

BenHenning reviewed (changes requested):
My initial comment overlaps with one of Sean's earlier comments.

However, before I do a full pass on this can you rebase this & #62 on top of the introduce-domain-module branch and change the PR's base branches to introduce-domain-module? That way you can remove the new module bits and we can review just the new code in the context of what #64 is checking in.

Commit:
Added shared project dictionary

Commit:
Added shared project dictionary

Commit:
Added shared project dictionary

Commit:
Added shared project dictionary

Commit:
Merge branch 'gae-abstraction-topic-page' of https://github.com/rt4914/oppia-android into gae-abstraction-topic-page

rt4914 commented:
> My initial comment overlaps with one of Sean's earlier comments.
> 
> However, before I do a full pass on this can you rebase this & #62 on top of the introduce-domain-module branch and change the PR's base branches to introduce-domain-module? That way you can remove the new module bits and we can review just the new code in the context of what #64 is checking in.

Hi Ben,
I have done the rebase and changed the PR's base branch for this PR only. Actually, I am using rebase for the first time and therefore I just want to confirm if its correct or not.

Also, if this PR is correct then probably you should just check my #69 PR because that contains all the API services and models and therefore finally if #69 PR is finalised then we can close other PRs for GAE abstraction.

I have checked all the Models and services with valid API calls, so they are working as intended.

Commit:
Null support for objects added

BenHenning committed:
Fix part of #4: Introduce domain module (#64)

* Migrate ListenableFuture to a Kotlin coroutine, and split HomeActivity
into both an activity and a fragment.

* Introduce the domain and testsupport modules.

The domain module includes a user app history controller that provides
instances of a new AsyncResult interface that's meant to be a potential
bridge between Kotlin coroutines and LiveData. The exact design of how
this should work needs to be determined as part of #6.

This also includes a new testsupport module that's required due to
https://github.com/robolectric/robolectric/pull/4736. This is supporting a
new test for the app history controller that leverages an AndroidX
activity scenario to test the live data.

Note that the test does not yet work since there's a race condition
between the LiveData's coroutines completing and the test continuing to
verify the state of the activity. This needs to be resolved, likely by
waiting for test visual elements to change based on the LiveData result.

Additional tests need to be added for other new components, and some
slight cleaning up may be necessary.

* Fix binary build: testsupport should only be depended on for tests, not
for production configurations.

* Further attempts to try and test the asynchronous behavior of the new
LiveData + kotlin coroutine system.

This introduces a custom LiveData to mediate between Kotlin coroutines
in a way that allows tests to actually ensure the coroutines resolve
quickly, and ensure lifecycle safety through standard LiveData
mechanisms.

However, this doesn't actually fix the underlying tests (which have been
changed in a number of ways in this commit in an effort to try and get
them to pass). The kotlinx coroutine deps were downgraded to work around
a NoSuchMethodError (see
https://github.com/Kotlin/kotlinx.coroutines/issues/1331). After
introducing a runBlockingTest structure and updating the user app
histoyr controller to use the test context, the operations seem to work
correctly. However, one job is hanging which is causing the test to
still fail (it appears to be deadlocked--no known amount of time allows
it to resolve).

This new NotifiableAsyncLiveData also appears to break the production
behavior, as well. The LiveData result is no longer being observed by
the data binding transformation function, so it seems the coroutines
aren't being executed, aren't completing, or something is deadlocking
somewhere. Further investigation is needed.

* Fix the app not properly binding to the live data by ensuring the data
binder had the view model set properly. Fix resource binding issues in
Robolectric tests by ensuring binary resources are provided to
Robolectric. Fix the controller tests not working by introducing a
temporary, custom CoroutineLiveData that ensures no long-living jobs
continue running after the live data is completed.

* Clean up tests (we no longer need a testsupport module for this module
introduction).

* Add test for AsyncResult. Set code style for Groovy files & clean up all
build gradle files. Add missing newlines at end of various files.

* Reformat top-level build.gradle file, as well.

* Address review comments: replace name-based TODOs with issues.

* Address review comments: add EOF new lines.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt:
```diff
@@ -0,0 +1,30 @@
+package org.oppia.data.backends.gae
+
+import okhttp3.*
```
Please expand imports instead of using wildcards. It helps avoid namespace pollution, and provides better context to readers on needed dependencies.

rt4914 replied:
Solved this

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt:
```diff
@@ -0,0 +1,30 @@
+
+import java.io.IOException
+
+class ModifyJsonInterceptor : Interceptor {
```
For here & all other new classes in this PR: please add Javadoc documentation for them.

rt4914 replied:
Added Javadoc to every class and also changed ModifyJsonInterceptor to NetworkInterceptor

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt:
```diff
@@ -0,0 +1,30 @@
+    val request = chain.request()
+    val response = chain.proceed(request)
+
+    if (response.code() == 200) {
```
Please move ``200`` to a module constant to provide context to it so that it's not a seemingly magic number. If Retrofit has a constant for it, that would be even better.

rt4914 replied:
Created new Constants file at module level and introduced 200 as a constant.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt:
```diff
@@ -0,0 +1,30 @@
+      if (response.body() != null) {
+        var rawJson = response.body()!!.string()
+        if (rawJson.startsWith(NetworkSettings.XSSI_PREFIX)) {
+          rawJson = rawJson.substring(rawJson.indexOf('\n') + 1)
```
Should we maybe just remove the XSSI_PREFIX instead? Extra whitespace should just be ignored by the JSON parser, and this guards against the assumption that a newline needs to be present for the prefix.

rt4914 replied:
Code updated and used removePrefix method.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt:
```diff
@@ -0,0 +1,30 @@
+        return response.newBuilder().body(body).build()
+      }
+    } else if (response.code() == 403) {
+      //TODO(#5): Identify how error handling will work (e.g. what Retrofit does),
```
Per our conversation last week, is there action needed to be taken here? I thought we were just going to rely on the default Retrofit behavior for error handling (e.g. that errors should be propagated to the Retrofit Call objects.

rt4914 replied:
Yes, this is not required now and has been removed from the code.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt:
```diff
@@ -0,0 +1,8 @@
+
+object NetworkSettings{
+
+  const val BASE_URL = "https://oppia.org"
```
Please document this & next comment.

Also, please add a TODO on #74 to move these to the DI graph.

rt4914 replied:
I have added a TODO comment. Not sure if you wanted me to do something else. So please check this once again.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt:
```diff
@@ -0,0 +1,8 @@
+package org.oppia.data.backends.gae
+
+object NetworkSettings{
```
Nit: need space before ``{``

Ditto for other files in this PR.

rt4914 replied:
Solved

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt:
```diff
@@ -0,0 +1,26 @@
+
+  val retrofitInstance: Retrofit?
+    get() {
+      if (retrofit == null) {
```
Rather than having a var store retrofit, could we use lazy initialization and keep the field as a value?

See: https://www.baeldung.com/kotlin-lazy-initialization.

rt4914 replied:
Solved. Made use of kotlin-lazy as per Ben's suggestion.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:
```diff
@@ -0,0 +1,25 @@
+
+  /*
+   * Ignore:
+   * is_moderator, is_admin, username, user_email, iframed, is_topic_manager, is_super_admin
```
Let's just omit this comment. If we aren't including these, I don't think we need to include them.

rt4914 replied:
Removed this comment.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:
```diff
@@ -0,0 +1,25 @@
+  @Json(name = "topic_name") val topicName: String?,
+  @Json(name = "canonical_story_dicts") val canonicalStoryDicts: List<CanonicalStorySummary?>?,
+  @Json(name = "additional_story_dicts") val additionalStoryDicts: List<AdditionalStorySummary?>?,
+  // skill_descriptions map has skill id as key and skill name as value
```
Please frame this as a Javadoc instead of a line comment. Ditto elsewhere for similar comments.

rt4914 replied:
Changed comments to Javadoc format

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:
```diff
@@ -0,0 +1,25 @@
+
+  @Json(name = "topic_id") val topicId: String?,
+  @Json(name = "topic_name") val topicName: String?,
+  @Json(name = "canonical_story_dicts") val canonicalStoryDicts: List<CanonicalStorySummary?>?,
```
For lists & collections of objects, can they actually be absent?

rt4914 replied:
I have seen responses in which lists and collections are also absent as well as null.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:
```diff
@@ -0,0 +1,25 @@
+import com.squareup.moshi.JsonClass
+
+@JsonClass(generateAdapter = true)
+data class Topic(
```
Please add Javadocs for these data objects (here & elsewhere). If possible, it'd be really useful to include links to the Python code that these structures correspond to for reference.

rt4914 replied:
Added Javadoc comment and python file links

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/SubtopicSummary.kt:
```diff
@@ -0,0 +1,13 @@
+@JsonClass(generateAdapter = true)
+data class SubtopicSummary(
+
+  @Json(name = "id") val id: String?,
```
Perhaps ``subtopicId`` might be clearer?

rt4914 replied:
Changed id to subtopicId

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/CanonicalStorySummary.kt:
```diff
@@ -0,0 +1,13 @@
+@JsonClass(generateAdapter = true)
+data class CanonicalStorySummary(
+
+  @Json(name = "id") val id: String?,
```
Perhaps ``storyId`` instead?

rt4914 replied:
Changed id to storyId

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/AdditionalStorySummary.kt:
```diff
@@ -0,0 +1,13 @@
+@JsonClass(generateAdapter = true)
+data class AdditionalStorySummary(
+
+  @Json(name = "id") val id: String?,
```
Are there any differences between this & the canonical version? Are they also different structures in the backend? It might be nice to consolidate them into 1 object if not, since they seem to be the same structure with different uses.

rt4914 replied:
Merged Additional and Canonical into "StorySummary" model.

BenHenning reviewed (changes requested):
Thanks @rt4914! There should be a PR coming out soon for the final module to wrap up  #4. We may block this until that PR is out to avoid introducing the data module twice.

BenHenning commented:
Actually one more request: @rt4914 can you also add tests for the interceptor? Since it's doing logic, we should have tests to ensure it continues to do the right thing in the long-term, if possible.

Commit:
Resolving merge confilts with domain module

Commit:
Resolving merge conflict in UserAppHistroyController

Commit:
Added javadocs and more clean code

Commit:
Optimised XSSI_PREFIX handling code

Commit:
Resolved merge conflict from HomeFragment

Commit:
Added python model structure links

Commit:
Linked models to python code

Commit:
NetworkInterceptor test cases added

Commit:
code cleaning

rt4914 commented:
> Actually one more request: @rt4914 can you also add tests for the interceptor? Since it's doing logic, we should have tests to ensure it continues to do the right thing in the long-term, if possible.

Hi @BenHenning , I have written test cases for NetworkInterceptor. I have taken help from your domain testing code and https://developer.android.com/training/testing/unit-testing/local-unit-tests

As this is me first time experience in writing a test cases please have a good look. Also, I have checked and both of these test cases pass.

Commit:
Added javadoc comment to removeXSSIPrefixFromResponse method

Commit:
code cleaning

seanlip commented:
Thanks @rt4914 -- no further comments from my end. I'll need to defer the rest of the review to @BenHenning.

rt4914 commented:
@BenHenning , I have created another PR for this on #75 because there were a lot of merge conflicts on this PR. I will close this PR once you merge #75 PR.  Also, just for clarification the new branch name is gae-abstraction-topic-page-2

BenHenning commented:
Closing this since it's been replaced by #78.
